app_utils/sections/experiment.py

```diff
@@ -1471,6 +1481,121 @@ async def experiment_download_model(q: Q, error: str = ""):
+        )
+        pretrained_model_dir = checkpoint_path
+        qmodel = AutoGPTQForCausalLM.from_pretrained(pretrained_model_dir, quantize_config)
+
+        def tokenize(examples):
+            instructions = examples["instruction"]
+            outputs = examples["output"]
+
+            input_ids = []
+            attention_mask = []
+
+            for raw_prompt, answer in zip(instructions, outputs):
+
+                full_prompt = cfg.dataset.dataset_class.parse_prompt(cfg, raw_prompt)
+                full_prompt_with_answer = f"{full_prompt}{answer}"
+                logger.info(f"Full prompt: {full_prompt}")
+                logger.info(f"Full prompt with answer: {full_prompt_with_answer}")
+
+                tokenized_data = tokenizer(full_prompt_with_answer)
+
+                input_ids.append(tokenized_data["input_ids"][: tokenizer.model_max_length])
+                attention_mask.append(tokenized_data["attention_mask"][: tokenizer.model_max_length])
+
+            return {
+                "input_ids": input_ids,
+                "attention_mask": attention_mask,
+            }
+
+        # FIXME: Extract from real dataset
+        raw_data = Dataset.from_dict({"instruction": [
+            "This is a test"
+        ], "input": [
+            "This is the answer"
+        ]})
+
+        n_samples = 128
+        # Sample from the dataset if n_samples is provided and less than the dataset size
+        if n_samples is not None and n_samples < len(raw_data):
+            raw_data = raw_data.shuffle(seed=42).select(range(n_samples))
+
+        examples = raw_data.map(
+            tokenize,
+            batched=True,
+            batch_size=len(raw_data),
+            num_proc=1,
+            keep_in_memory=True,
+            load_from_cache_file=False,
+            remove_columns=["instruction", "input"]
+        )
+
+        # Convert to PyTorch tensors
+        examples.set_format(type='torch', columns=['input_ids', 'attention_mask'])
+
+        examples_for_quant = [
+            {"input_ids": example["input_ids"], "attention_mask": example["attention_mask"]}
+            for example in examples
+        ]
+
+        logger.info("Quantize...")
+        start = time.time()
+        # quantize model, the examples should be list of dict whose keys can only be "input_ids" and "attention_mask"
+        qmodel.quantize(
+            examples_for_quant,
+            batch_size=1,
+            # batch_size=args.quant_batch_size,
+            # use_triton=args.use_triton,
+            # autotune_warmup_after_quantized=args.use_triton
+        )
+        end = time.time()
+        logger.info(f"quantization took: {end - start: .4f}s")
+
+        # save quantized model
+        logger.info("Saving quantized model...")
+        quantized_model_dir = checkpoint_path + "_quantized"
+        qmodel.save_quantized(quantized_model_dir, use_safetensors=True)
```

Comment:
Need a way to push after saving. Maybe do it in one step, as AutoGPT states is possible all together?